Bottom: 7cfa8b8bbb7449eac973ab817f043b3f1a3850b7
Top:    7bf18695fef78ce1cf020b98bda96e2687917519
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-08-07 09:37:24 -0700

rspreload: Add fstat support

vsftpd calls fstat on a socket.  Fake it out.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>


---

diff --git a/src/preload.c b/src/preload.c
index b18d310..33eb9a3 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -84,6 +84,7 @@ struct socket_calls {
 			  void *optval, socklen_t *optlen);
 	int (*fcntl)(int socket, int cmd, ... /* arg */);
 	int (*dup2)(int oldfd, int newfd);
+	int (*fstat)(int fd, struct stat *buf);
 };
 
 static struct socket_calls real;
@@ -259,6 +260,7 @@ static void init_preload(void)
 	real.getsockopt = dlsym(RTLD_NEXT, "getsockopt");
 	real.fcntl = dlsym(RTLD_NEXT, "fcntl");
 	real.dup2 = dlsym(RTLD_NEXT, "dup2");
+	real.fstat = dlsym(RTLD_NEXT, "fstat");
 
 	rs.socket = dlsym(RTLD_DEFAULT, "rsocket");
 	rs.bind = dlsym(RTLD_DEFAULT, "rbind");
@@ -961,3 +963,17 @@ int dup2(int oldfd, int newfd)
 	atomic_inc(&oldfdi->refcnt);
 	return newfd;
 }
+
+int fstat(int socket, struct stat *buf)
+{
+	int fd, ret;
+
+	if (fd_get(socket, &fd) == fd_rsocket) {
+		ret = real.fstat(socket, buf);
+		if (!ret)
+			buf->st_mode = (buf->st_mode & ~S_IFMT) | __S_IFSOCK;
+	} else {
+		ret = real.fstat(fd, buf);
+	}
+	return ret;
+}
